src/object.c

```diff
@@ -135,6 +150,148 @@ robj *createStringObject(const char *ptr, size_t len) {
+        data += 1 + hdr_size;
+        return (sds)data;
+    }
+    return NULL;
+}
+
+long long valkeyGetExpire(const valkey *val) {
+    unsigned char *data = (void *)(val + 1);
+    if (val->hasexpire) {
+        return *(long long *)data;
+    } else {
+        return -1;
+    }
+}
+
+void valkeySetExpire(valkey *val, long long expire) {
+    unsigned char *data = (void *)(val + 1);
+    assert(val->hasexpire);
+    *(long long *)data = expire;
+}
+
+/* Attaches a key to the object, without reallocating the object. */
+static void objectSetKey(robj *val, const sds key) {
+    assert(val->hasembkeyptr && !val->hasembkey && valkeyGetKey(val) == NULL);
+
+    /* Find the correct location in val's data field. */
+    unsigned char *data = (void *)(val + 1);
+    if (val->hasexpire) {
+        /* Skip expire field */
+        data += sizeof(long long);
+    }
+    sds oldkey = *(sds *)data;
+    if (oldkey != NULL) sdsfree(oldkey);
+    *(sds *)data = sdsdup(key);
+}
+
+/* Converts (updates, possibly reallocates) 'val' to a valkey object by
+ * attaching a key to it. This functions takes ownership of "one refcount" of
+ * val. Think of val's refcount being decremented by one and the returned
+ * object's refcount being incremented by one. Confused? Simply use the returned
+ * object instead of 'val' after calling this function and you'll be fine. */
+valkey *objectConvertToValkey(robj *val, const sds key) {
+    /* If a key pointer is already embedded, free that sds string first. */
+    if (val->hasembkeyptr) {
+        /* Find the correct location in val's data field. */
+        unsigned char *data = (void *)(val + 1);
+        if (val->hasexpire) {
+            /* Skip expire field */
+            data += sizeof(long long);
+        }
+        sds oldkey = *(sds *)data;
+        sdsfree(oldkey);
+        *(sds *)data = NULL;
+    }
+
+    if (val->encoding == OBJ_ENCODING_EMBSTR) {
+        /* Create a new object with the key embedded and return it. */
+
+        /* TODO: If there's space in val's allocation, we can embed the key
+         * there and memmove the the embedded value, without creating a new
+         * object.
+         *
+         * TODO: If key + value are too large (allocation > 64 bytes) we may not
+         * want to embed both of them. We can embed one or the other depending
+         * on sizes. */
+
+        /* Create a new object with val and key embedded and decrement the
+         * reference counter of 'val'. */
+
+        /* Calculate sizes */
+        size_t key_sds_size = sdscopytobuffer(NULL, 0, key, NULL);
+        size_t val_len = sdslen(val->ptr);
+
+        size_t min_size = sizeof(robj);
+        min_size += sizeof(long long); /* expire */
+        /* Size of embedded key, incl. 1 byte for prefixed sds hdr size. */
+        min_size += 1 + key_sds_size;
+        /* Size of embedded value (EMBSTR) including \0 term. */
+        min_size += sizeof(struct sdshdr8) + val_len + 1;
```

Comment:
EMBSTR was always hard-coded to using sdshdr8. This is old code, refactored. I wanted to avoid rewriting it.

In the future, a good thing would be to get rid of the `ptr` field in robj and instead compute it using a function. Then we can save another 8 bytes for embedded strings.